src/main/java/bio/terra/dao/DatasetTableDao.java

```diff
@@ -1,17 +1,90 @@
 package bio.terra.dao;
 
+import bio.terra.configuration.DataRepoJdbcConfiguration;
+import bio.terra.metadata.Column;
+import bio.terra.metadata.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
 import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
 import org.springframework.stereotype.Repository;
 
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
 @Repository
 public class DatasetTableDao extends TableDaoBase {
 
+    private static final Logger logger = LoggerFactory.getLogger(DatasetTableDao.class);
+
+    private static final String sqlInsertTable = "INSERT INTO dataset_table " +
+        "(name, dataset_id, primary_key) VALUES (:name, :dataset_id, :primary_key)";
+
+    private final DataSource jdbcDataSource;
     private final NamedParameterJdbcTemplate jdbcTemplate;
 
+
     @Autowired
-    public DatasetTableDao(NamedParameterJdbcTemplate jdbcTemplate) {
-        super(jdbcTemplate, "dataset_table", "dataset_column", "dataset_id");
-        this.jdbcTemplate = jdbcTemplate;
+    public DatasetTableDao(DataRepoJdbcConfiguration dataRepoJdbcConfiguration) {
+        super(dataRepoJdbcConfiguration, "dataset_table", "dataset_column", "dataset_id");
+        this.jdbcDataSource = dataRepoJdbcConfiguration.getDataSource();
+        this.jdbcTemplate = new NamedParameterJdbcTemplate(this.jdbcDataSource);
+    }
+
+    @Override
+    protected UUID createTable(UUID parentId, Table table) {
+        MapSqlParameterSource params = new MapSqlParameterSource();
+        DaoKeyHolder keyHolder = new DaoKeyHolder();
+
+        logger.info(parentId.toString());
+        params.addValue("name", table.getName());
+        params.addValue("dataset_id", parentId);
+
+        List<String> naturalKeyStringList = table.getPrimaryKey()
+            .orElseThrow(() -> new IllegalStateException("Dataset table " + table.getName() + " has no primary key"))
```

Comment:
I don't think that should be an error. There are tables that have no primary keys. I don't think we should make it illegal to represent data that way.

Instead, I'd suggest that you can't do an upsert on a table that doesn't have a primary key defined.

Reply:
Not a hill I'm willing to die on, but just to explain my thoughts...

There are three possible states here:
1. `Optional.EMPTY`
2.  A non-empty `Optional` containing an empty list
3. A non-empty `Optional` containing a non-empty list

Of the three, 1. is always the case for snapshot `Table`s (there's no primary key column) and never the case for dataset `Table`s (the primary key column is non-nullable). Since this is the `DatasetTableDao`, seeing an empty `Optional` means something has gone terribly wrong.

So, I think there's some value to the exception here. If I've convinced you, I can add a comment with that explanation. Otherwise I can revert back to `Collections.emptyList()`

Reply:
I was misled by the error message. The error isn't that a primary key is missing; I think it should be CorruptMetadataException: we have a state that should be invalid; our code is broken.

I think it would be simpler to make the rule - enforced in Table class - that only two states are allowed to exist:
1. Null - no primary key
2. Non-empty list - primary key defined
If you get Empty list, it is an error. And it should not be constructable by Table.

Wrapped in Optional if you like that sort of thing.